Initial credential provider work

nkolev92 opened the pull request:
## Bug
Fixes: https://github.com/NuGet/Home/issues/6699
Regression: No  
If Regression then when did it last work:   
If Regression then how are we preventing it in future:   

## Fix
Details: 
This is targeting a feature branch. 

Increment the plugin version. 
Refactor plugin resource provider to plugin manager (most of the code is verbatim. The code path for the download plugin is kept the same)
Add securecredentialprovider builder + secure credential provider. 
Load extensionsin msbuild and nuget.exe. 
Add a new operation claim and new messages as required

As this is a first part in a series of items...here are the known issues not a target of this PR. 
- [ ] "Discovering" the plugins to create credential providers is costly, mostly because currently the plugin discoverer does the verification of the plugin. 
This will be addressed in a further work item. https://github.com/NuGet/Home/issues/6744
- [ ] No capabilities to discover plugins in any special way. https://github.com/NuGet/Home/issues/6699
- [ ] The tools that patch dotnet.exe are broken due to the dependency change where credentials is now a part of the dlls required to launch dotnet.exe. 
This will be addressed in dev initially, and it will be fixed, when we have a version of dotnet.exe built with nuget.credentials in. 

## Testing/Validation
Tests Added: Yes
Reason for not adding tests:  
Validation done:  
Automated tests + Manual testing

//cc 
@jeffkl Once this is merged, you should be able to use the feature branch safely for your testing.

nkolev92 commented:
ping @NuGet/nuget-client

nkolev92 committed:
rev up plugin version, add plugin manager, add secure credential plugin provider, add a default provider, load these extensions in msbuild and nuget.exe, add new auth capabilities

dtivel commented on src/NuGet.Clients/NuGet.Credentials/SecureCredentialProviderBuilder.cs:
```diff
@@ -0,0 +1,45 @@
+
+        public SecureCredentialProviderBuilder(Common.ILogger logger)
+        {
+            _logger = logger;
```
Can `logger` be `null`?

dtivel commented on src/NuGet.Clients/NuGet.Credentials/SecurePluginCredentialProvider.cs:
```diff
@@ -0,0 +1,119 @@
+        private Common.ILogger Logger { get; set; }
+
+        // We use this to avoid needlessly instantiating plugins if they don't support authentication.
+        private bool IsAnAuthenticationPlugin { get; set; } = true;
```
You don't need private properties.  Make these all private readonly fields.

dtivel commented on src/NuGet.Clients/NuGet.Credentials/SecurePluginCredentialProvider.cs:
```diff
@@ -0,0 +1,119 @@
+            }
+            if (pluginDiscoveryResult.PluginFile.State != PluginFileState.Valid)
+            {
+                throw new ArgumentException("Cannot create a provider from an invalid plugin. Plugin state: " + pluginDiscoveryResult.PluginFile.State + " " + pluginDiscoveryResult.Message);
```
Use a string resource.

dtivel replied:
Add an argument to `ArgumentException` constructor:  `nameof(pluginDiscoveryResult)`

dtivel commented on src/NuGet.Clients/NuGet.Credentials/SecurePluginCredentialProvider.cs:
```diff
@@ -0,0 +1,119 @@
+        /// <param name="message">A message provided by NuGet to show to the user when prompting.</param>
+        /// <param name="nonInteractive">If true, the plugin must not prompt for credentials.</param>
+        /// <param name="cancellationToken">A cancellation token.</param>
+        /// <returns>A credential object.  If </returns>
```
Either finish the sentence starting with "If" or delete it.

dtivel commented on src/NuGet.Clients/NuGet.Credentials/SecurePluginCredentialProvider.cs:
```diff
@@ -0,0 +1,119 @@
+            return taskResponse;
+        }
+
+        private static CredentialResponse GetCredentialResponseToCredentiaResponse(GetAuthenticationCredentialsResponse credentialResponse)
```
Credential is misspelled.

dtivel commented on src/NuGet.Clients/NuGet.Credentials/SecurePluginCredentialProvider.cs:
```diff
@@ -0,0 +1,119 @@
+using NuGet.Configuration;
+using NuGet.Protocol.Core.Types;
+using NuGet.Protocol.Plugins;
+using System.Linq;
```
Sort

dtivel commented on src/NuGet.Core/NuGet.Commands/RestoreCommand/RestoreRunner.cs:
```diff
@@ -41,7 +43,34 @@ public static class RestoreRunner
+        }
+
+        // Add only the secure plugin. This will be done when there's nothing set
+        private static async Task<IEnumerable<Credentials.ICredentialProvider>> GetCredentialProviders(ILogger logger )
```
`GetCredentialProvidersAsync(...)`

dtivel commented on src/NuGet.Core/NuGet.Commands/RestoreCommand/RestoreRunner.cs:
```diff
@@ -41,7 +43,34 @@ public static class RestoreRunner
         }
 
-        /// <summary>
+        private static async Task SetDefaultCredentialProvider(ILogger logger)
```
`SetDefaultCredentialProviderAsync(...)`

dtivel commented on src/NuGet.Core/NuGet.Protocol/Plugins/PluginFactory.cs:
```diff
@@ -137,6 +137,19 @@ public void Dispose()
+            _plugins.TryRemove(plugin.FilePath, out val);
+            if (val.IsValueCreated)
+            {
+                if (val.Value.IsCompleted)  // nothing to dispose of if completed
```
// nothing to dispose of if NOT completed

Just remove the comment though.

dtivel commented on src/NuGet.Core/NuGet.Protocol/Plugins/PluginFactory.cs:
```diff
@@ -137,6 +137,19 @@ public void Dispose()
+        {
+            Lazy<Task<IPlugin>> val;
+            _plugins.TryRemove(plugin.FilePath, out val);
+            if (val.IsValueCreated)
```
if (_plugins.TryRemove(plugin.FilePath, out val) && val.IsValueCreated)

dtivel commented on src/NuGet.Core/NuGet.Protocol/Plugins/PluginFactory.cs:
```diff
@@ -137,6 +137,19 @@ public void Dispose()
             return new NoOpDisposePlugin(lazyTask.Value.Result);
         }
 
+        public void DisposePlugin(IPlugin plugin)
```
Check that `plugin` is not `null` and add XML doc comments.

dtivel replied:
I think this method should be removed or at least removed from the interface and made internal.

dtivel commented on src/NuGet.Core/NuGet.Protocol/Plugins/Messages/GetAuthenticationCredentialsRequest.cs:
```diff
@@ -0,0 +1,30 @@
+{
+
+
+    public class GetAuthenticationCredentialsRequest
```
`sealed`

dtivel commented on src/NuGet.Core/NuGet.Protocol/Plugins/Messages/GetAuthenticationCredentialsRequest.cs:
```diff
@@ -0,0 +1,30 @@
+        public bool IsRetry { get; }
+
+        [JsonRequired]
+        public bool NonInteractive { get; }
```
How about `IsNonInteractive` (parameter name too)?  Ideally, bool members begin with `Is`, `Has`, `Can`, etc.

dtivel commented on src/NuGet.Core/NuGet.Protocol/Plugins/Messages/GetAuthenticationCredentialsResponse.cs:
```diff
@@ -0,0 +1,60 @@
+
+namespace NuGet.Protocol.Plugins
+{
+    public class GetAuthenticationCredentialsResponse
```
`sealed`

dtivel commented on src/NuGet.Core/NuGet.Protocol/Plugins/PluginManager.cs:
```diff
@@ -0,0 +1,348 @@
+            }
+        }
+
+        public Task DisposeOfPlugin(IPlugin plugin)
```
This is odd.

First, it should be `DisposePlugin(...)` to be consistent with `IPluginFactory.DisposePlugin(...)`.

Second, `IPlugin` already implements `IDisposable`.  If I wanted to dispose an `IPlugin` instance, I'd just call `IPlugin.Dispose()`.  I think a cleaner design would make it so that when you call `IPlugin.Dispose()` it will internally unregister with the `PluginManager`.

dtivel replied:
I don't think this is necessary.

dtivel commented on src/NuGet.Core/NuGet.Protocol/Plugins/PluginManager.cs:
```diff
@@ -0,0 +1,348 @@
+            return Task.CompletedTask;
+        }
+
+        public async Task<IEnumerable<PluginDiscoveryResult>> FindAvailablePlugins(CancellationToken cancellationToken)
```
`FindAvailablePluginsAsync(...)`

dtivel commented on src/NuGet.Core/NuGet.Protocol/Plugins/PluginManager.cs:
```diff
@@ -0,0 +1,348 @@
+            return await _discoverer.Value.DiscoverAsync(cancellationToken);
+        }
+
+        public async Task<IEnumerable<PluginCreationResult>> TryCreate(
```
`TryCreateAsync(...)`

dtivel commented on src/NuGet.Core/NuGet.Protocol/Plugins/PluginManager.cs:
```diff
@@ -0,0 +1,348 @@
+        }
+
+
+        private async Task<Lazy<IPluginMulticlientUtilities>> PerformOneTimePluginInitialization(IPlugin plugin, CancellationToken cancellationToken)
```
`PerformOneTimePluginInitializationAsync(...)`

dtivel commented on src/NuGet.Core/NuGet.Protocol/Plugins/ProtocolConstants.cs:
```diff
@@ -14,7 +14,12 @@ public static class ProtocolConstants
+        public static readonly SemanticVersion CurrentVersion = new SemanticVersion(major: 2, minor: 0, patch: 0);
+
+        /// <summary>
+        /// The current protocol version.
```
Update XML doc comments.  I think this is the minimum supported protocol version.

dtivel commented on test/NuGet.Clients.Tests/NuGet.Credentials.Test/SecurePluginCredentialProviderBuilderTests.cs:
```diff
@@ -0,0 +1,119 @@
+        private const string _pluginHandshakeTimeoutEnvironmentVariable = "NUGET_PLUGIN_HANDSHAKE_TIMEOUT_IN_SECONDS";
+        private const string _pluginIdleTimeoutEnvironmentVariable = "NUGET_PLUGIN_IDLE_TIMEOUT_IN_SECONDS";
+        private const string _pluginPathsEnvironmentVariable = "NUGET_PLUGIN_PATHS";
+        private const string _pluginRequestTimeoutEnvironmentVariable = "NUGET_PLUGIN_REQUEST_TIMEOUT_IN_SECONDS";
```
These environment variables are duplicated [here](https://github.com/NuGet/NuGet.Client/pull/2126/files#diff-8906b0502ac7cca79677892dd57c8193R65).  Extract them into test constants.

dtivel commented on test/NuGet.Clients.Tests/NuGet.Credentials.Test/PluginManagerMock.cs:
```diff
@@ -0,0 +1,234 @@
+                            PluginFileState pluginFileState,
+                TestExpectation expectations
+            )
+        {
```
There's some crazy indention going on throughout this file.  Try formatting document and fixing up places.

dtivel commented on src/NuGet.Clients/NuGet.Credentials/SecurePluginCredentialProvider.cs:
```diff
@@ -0,0 +1,119 @@
+
+namespace NuGet.Credentials
+{
+    public class SecurePluginCredentialProvider : ICredentialProvider
```
Either add XML doc comments to the type and all public members or don't any at all.  Two members have XML doc comments already.

Also, can be `sealed`?

dtivel commented on src/NuGet.Clients/NuGet.Credentials/SecureCredentialProviderBuilder.cs:
```diff
@@ -0,0 +1,45 @@
+
+namespace NuGet.Credentials
+{
+    public class SecureCredentialProviderBuilder
```
Can be `sealed`?

dtivel commented on src/NuGet.Core/NuGet.Protocol/Plugins/MessageMethod.cs:
```diff
@@ -91,6 +91,11 @@ public enum MessageMethod
+        SetLogLevel,
+
+        /// <summary>
+        /// Get operation claims
```
Update XML doc comments.

dtivel commented on src/NuGet.Core/NuGet.Protocol/Plugins/MessageMethod.cs:
```diff
@@ -91,6 +91,11 @@ public enum MessageMethod
+        /// <summary>
+        /// Get operation claims
+        /// </summary>
+        GetAuthCredentials,
```
I'd prefer that you not abbreviate here, especially since it could be authentication or authorization.

dtivel commented on src/NuGet.Core/NuGet.Protocol/Plugins/Messages/GetAuthenticationCredentialsResponse.cs:
```diff
@@ -0,0 +1,60 @@
+        /// Gets or sets the list of authentication types this credential is applicable to. Useful values include
+        /// <c>basic</c>, <c>digest</c>, <c>negotiate</c>, and <c>ntlm</c>
+        /// </summary>
+        public IList<string> AuthTypes { get; }
```
Do not abbreviate.

dtivel replied:
Add XML doc comments to the class and all its public members or do not add any at all.

dtivel commented on src/NuGet.Core/NuGet.Protocol/Plugins/Messages/GetAuthenticationCredentialsResponse.cs:
```diff
@@ -0,0 +1,60 @@
+
+
+        [JsonConstructor]
+        public GetAuthenticationCredentialsResponse(string username, string password, string message, IList<string> authTypes, MessageResponseCode responseCode)
```
Do not abbreviate `authTypes`.

dtivel commented on src/NuGet.Core/NuGet.Protocol/Plugins/IPluginFactory.cs:
```diff
@@ -42,5 +42,11 @@ public interface IPluginFactory : IDisposable
+        /// Dispose of that plugin if possible
+        /// </summary>
+        /// <param name="plugin"></param>
+        void DisposePlugin(IPlugin plugin);
```
Undo this change.

dtivel commented on test/NuGet.Clients.Tests/NuGet.Credentials.Test/PluginManagerMock.cs:
```diff
@@ -0,0 +1,234 @@
+        public SemanticVersion PluginVersion { get; }
+        public Uri Uri { get; }
+        public string AuthUsername { get; }
+        public string AuthPassword { get; }
```
Un-abbreviate `Auth`.

dtivel commented on test/NuGet.Clients.Tests/NuGet.Credentials.Test/SecurePluginCredentialProviderTests.cs:
```diff
@@ -0,0 +1,198 @@
+        [Fact]
+        public void Create_ThrowsForInvalidPlugin()
+        {
+            var exception = Assert.Throws<ArgumentException>(
```
Validate the return value (e.g.:  check parameter name) or don't bother capturing it in the first place.

dtivel commented on test/NuGet.Clients.Tests/NuGet.Credentials.Test/SecurePluginCredentialProviderTests.cs:
```diff
@@ -0,0 +1,198 @@
+                authUsername,
+                authPassword,
+                true,
+                true
```
Add named parameters to all these pesky constant literal arguments.  `null` and `true` aren't very readable.  This feedback applies to subsequent tests as well.

dtivel commented on test/NuGet.Core.Tests/NuGet.Protocol.Tests/Plugins/Messages/GetAuthenticationCredentialsRequestTests.cs:
```diff
@@ -0,0 +1,27 @@
+                () => new GetAuthenticationCredentialsRequest(
+                    uri,
+                    false,
+                    false
```
Named parameters will improve readability.

dtivel commented on test/NuGet.Clients.Tests/NuGet.Credentials.Test/SecurePluginCredentialProviderBuilderTests.cs:
```diff
@@ -0,0 +1,119 @@
+
+namespace NuGet.Credentials.Test
+{
+    public class SecurePluginCredentialProviderBuilderTests
```
You have a problem.

You have different tests classes partying on `PluginManager.Instance` potentially in parallel.  Some tests call [`Reinitialize(...)`](https://github.com/NuGet/NuGet.Client/pull/2126/files#diff-c67d2e9114567b403170c640b29bb7a3R99) and some tests call [`Dispose()`](https://github.com/NuGet/NuGet.Client/pull/2126/files#diff-3eddcf79cd5a584a191f9f68d2d1d61aR453).  This will lead to tests interfering with one another and causing seemingly random failures.

One solution would to make `PluginManager`'s constructor public for testing purposes and eliminate the static stuff (except for `Instance`).  The product will still use `PluginManager.Instance` but tests can new up a new `PluginManager` instance to test in total isolation from other tests.

You may even be able to get rid of `Reinitialize(...)` with this approach.

nkolev92 replied:
I couldn't quite get rid of reinitialize, but I'm now only using instance from production code during intiialization, and I pass down a plugin manager everywhere it's needed. 

I did face a couple of intermittent issues that I did brush off as flaky, but in the multiple runs after adding my change, they are yet to fail!

dtivel commented on src/NuGet.Core/NuGet.Protocol/Plugins/PluginManager.cs:
```diff
@@ -0,0 +1,348 @@
+                cancellationToken);
+            if (response == null)
+            {
+                return new List<OperationClaim>();
```
Will this work instead?

    return Array.Empty<OperationClaim>();

nkolev92 committed:
address style/readability feedback

nkolev92 committed:
Make the constructor of the pLuginManager singleton public for testing purposes

nkolev92 commented:
🔔 
@dtivel 
I have addressed your feedback. 

It's in 3 commits.
- stylistic changes
- re-use of singleton in tests cleanup
- disposing of plugins API cleanup.

@NuGet/nuget-client 
can someone else take a look as well?

nkolev92 commented:
ping @NuGet/nuget-client  @dtivel

dtivel commented on src/NuGet.Clients/NuGet.Credentials/SecurePluginCredentialProvider.cs:
```diff
@@ -0,0 +1,140 @@
+        /// <summary>
+        /// Create a credential provider based on provided plugin
+        /// </summary>
+        /// <param name="pluginDiscoveryResult"></param>
```
Missing `pluginManager` parameter.

dtivel commented on src/NuGet.Core/NuGet.Protocol/Plugins/Messages/GetAuthenticationCredentialsResponse.cs:
```diff
@@ -0,0 +1,80 @@
+        public string Password { get; }
+
+        /// <summary>
+        /// message - optional, can be used as a way to communicate to NuGet why the authentication faled.
```
failed

dtivel commented on src/NuGet.Core/NuGet.Protocol/Plugins/Messages/GetAuthenticationCredentialsResponse.cs:
```diff
@@ -0,0 +1,80 @@
+        /// Gets or sets the list of authentication types this credential is applicable to. Useful values include
+        /// <c>basic</c>, <c>digest</c>, <c>negotiate</c>, and <c>ntlm</c>
+        /// </summary>
+        public IList<string> AuthenticationTypes { get; }
```
`IReadOnlyList<string>`

nkolev92 replied:
While that's more explicit, I'd prefer to keep this as it matches the rest of the APIs for credentials in NuGet.
I'd like to avoid frequent IReadOnlyList <=> IList conversions. 

If you think otherwise, I'd be happy to create a quick task to rework the credential APIs as well (not sure if it's possible, but I don't think there's a blocker preventing us from doing it)

dtivel commented on src/NuGet.Core/NuGet.Protocol/Plugins/Messages/GetAuthenticationCredentialsResponse.cs:
```diff
@@ -0,0 +1,80 @@
+        /// <param name="responseCode"></param>
+        /// <exception cref="ArgumentException">If MessageResponseCode is not defined on this runtime</exception>
+        [JsonConstructor]
+        public GetAuthenticationCredentialsResponse(string username, string password, string message, IList<string> authenticationTypes, MessageResponseCode responseCode)
```
`IEnumerable<string> authenticationTypes`

nkolev92 replied:
Same as the above answer, the types need to be assignable.

dtivel commented on src/NuGet.Core/NuGet.Protocol/Providers/PluginResourceProvider.cs:
```diff
@@ -21,57 +21,20 @@ namespace NuGet.Protocol.Core.Types
-        private ConcurrentDictionary<PluginPackageSourceKey, Lazy<Task<IReadOnlyList<OperationClaim>>>> _pluginOperationClaims;
-        private ConcurrentDictionary<string, Lazy<IPluginMulticlientUtilities>> _pluginUtilities;
-        private string _rawPluginPaths;
+        private readonly PluginManager _pluginManager;
```
You removed `IDisposable` from `PluginResourceProvider`'s base list, but the class owns `_pluginManager`, which implements `IDisposable`.

nkolev92 replied:
The PluginManager is now singleton. 
IDisposable doesn't work here. 
I'll get back to you on this.

nkolev92 replied:
I've made the changes as per our offline discussion. 
IPluginManager is now referenced everywhere in product code to hide that PluginManager is disposable.
Cleaning up some extra logic before deeming this as ready for review again.

dtivel commented on src/NuGet.Core/NuGet.Protocol/Plugins/SmartDisposablePlugin.cs:
```diff
@@ -63,26 +63,24 @@ public sealed class NoOpDisposePlugin : IPlugin
         public void Dispose()
         {
+            var deregister = _deregister(_plugin);
+            _plugin.Dispose();
```
This absolutely won't work.  Let's go back to `NoOpDisposePlugin`.

dtivel commented on test/NuGet.Clients.Tests/NuGet.Credentials.Test/CredentialTestConstants.cs:
```diff
@@ -0,0 +1,13 @@
+
+namespace NuGet.Credentials.Test
+{
+    class CredentialTestConstants
```
internal static

jeffkl commented on src/NuGet.Clients/NuGet.Credentials/SecurePluginCredentialProvider.cs:
```diff
@@ -0,0 +1,134 @@
+        /// <param name="nonInteractive">If true, the plugin must not prompt for credentials.</param>
+        /// <param name="cancellationToken">A cancellation token.</param>
+        /// <returns>A credential object.</returns>
+        public async Task<CredentialResponse> GetAsync(Uri uri, IWebProxy proxy, CredentialRequestType type, string message, bool isRetry, bool nonInteractive, CancellationToken cancellationToken)
```
It appears that the LogRequestHandler is not wired up so I can't log things in my plugin.  Is that expected?

https://github.com/NuGet/NuGet.Client/blob/870340673ba75a95407ab55ce1f46f57fe9bf25b/src/NuGet.Core/NuGet.Protocol/Resources/DownloadResourcePlugin.cs#L139

nkolev92 replied:
@jeffkl 
It's done consciously right now. 
There's no logger setup for the credential plugin. 
The above approach is a little bit spotty, because there's no contract on when a plugin can log. 

I've created a task to follow up on this:
https://github.com/NuGet/Home/issues/6762

jainaashish replied:
Would it be possible to link MSBuildLogger with credential plugin? since that's already been passed to this provider.

nkolev92 replied:
@jainaashish 
Yeah, I think that'll be possible. 
The logger is propagated from RestoreTask -> RestoreRunner -> CredentialProviderBuilder -> CredentialProvider.

nkolev92 committed:
remove explicit dispose methods from the plugin manager

nkolev92 committed:
Extract IPLuginManager to hide that PluginManager is disposable

nkolev92 committed:
cleanup, add comments

nkolev92 commented:
@dtivel 
Ready for another look. 

//cc @NuGet/nuget-client

dtivel commented on src/NuGet.Core/NuGet.Protocol/Plugins/PluginManager.cs:
```diff
@@ -0,0 +1,368 @@
+            return utilities;
+        }
+
+        private void Reinitialize(IEnvironmentVariableReader reader,
```
Rename to `Initialize` since it's only called once per instance.

nkolev92 replied:
done.

dtivel commented on src/NuGet.Core/NuGet.Protocol/Plugins/PluginManager.cs:
```diff
@@ -0,0 +1,368 @@
+    /// A plugin manager. This manages all the live plugins and their operation claims.
+    /// Invoked in by both the credential provider and the PluginResourceProvider
+    /// </summary>
+    /// <remarks>This is unsealed only to facilitate testing.</remarks>
```
Is this still true?  If no, remove and mark class `sealed`.

nkolev92 replied:
marked it as sealed.

dtivel reviewed (approved):
It would be good to get another set of eyes on this, but looks good to me.

jainaashish commented on src/NuGet.Clients/NuGet.CommandLine/Commands/Command.cs:
```diff
@@ -189,11 +189,14 @@ private IEnumerable<NuGet.Credentials.ICredentialProvider> GetCredentialProvider
             var pluginProviders = new PluginCredentialProviderBuilder(extensionLocator, Settings, Console)
                 .BuildAll(Verbosity.ToString())
                 .ToList();
+            var securePluginProviders =  (new SecureCredentialProviderBuilder(PluginManager.Instance, Console)).BuildAll().Result;
```
use `await` instead of `Result` since `Result` has potential for deadlock if Task has not completed yet.

jainaashish commented on src/NuGet.Core/NuGet.Commands/RestoreCommand/RestoreRunner.cs:
```diff
@@ -41,7 +44,34 @@ public static class RestoreRunner
+            {
+                if (PreviewFeatureSettings.DefaultCredentialsAfterCredentialProviders)
+                {
+                    providers.Add(new DefaultCredentialsCredentialProvider());
```
this `DefaultCredentialsCredentialProvider` is so confusing... I know it's existing impl but shouldn't we just change it to something more meaningful like `DefaultNetworkCredentialProvider`

nkolev92 replied:
I'll change it in a subsequent PR.

jainaashish commented on src/NuGet.Core/NuGet.Protocol/Plugins/IPluginManager.cs:
```diff
@@ -0,0 +1,42 @@
+        /// <param name="cancellationToken"></param>
+        /// <exception cref="ArgumentNullException">Throw if <paramref name="source"/> is null </exception>
+        /// <returns>A PluginCreationResult</returns>
+        Task<IEnumerable<PluginCreationResult>> TryCreateAsync(
```
Should it just be `CreateAsync` since it doesn't return bool to tell if it was able to create or not. Eventually consumers have to rely on the null check. 

Or keep it `TryCreateAsync` but then return Task<bool> and out param for list of PluginCreationResult

nkolev92 replied:
updated

jainaashish commented on src/NuGet.Core/NuGet.Protocol/Plugins/Messages/GetAuthenticationCredentialsResponse.cs:
```diff
@@ -0,0 +1,83 @@
+        /// one element
+        /// </remarks>
+        public bool IsValid => ResponseCode == MessageResponseCode.Success && (!string.IsNullOrWhiteSpace(Username) || !string.IsNullOrWhiteSpace(Password))
+                               && (AuthenticationTypes == null || AuthenticationTypes.Any());
```
`(AuthenticationTypes == null || AuthenticationTypes.Any())` didn't understand this condition clearly... we dont accept when someone explicitly set this list to empty but fine with null?

nkolev92 replied:
Yeah, that's exactly what it's supposed to be. 
This is equivalent to our current authentication code. 
If it's not set we create NetworkCredential, if it is, we create AuthTypeFilteredCredentials.

jainaashish replied:
It just feels odd, ideally we should just take empty as null as well in this case.

jainaashish commented on src/NuGet.Core/NuGet.Protocol/Plugins/Messages/GetOperationClaimsRequest.cs:
```diff
@@ -9,42 +9,29 @@ namespace NuGet.Protocol.Plugins
-        /// <exception cref="ArgumentException">Thrown if <paramref name="packageSourceRepository" /> is either
-        /// <c>null</c> or an empty string.</exception>
-        /// <exception cref="ArgumentNullException">Thrown if <paramref name="serviceIndex" /> is <c>null</c>.</exception>
+        /// <remarks>Both packageSourceRepository and service index can be null. If they are, the operation claims request is considered as source agnostic</remarks>
```
`request is considered as source agnostic` what does this really mean?

nkolev92 replied:
GetOperationClaimsRequest takes a source..basically the question is:
- Which operations can you handle for this source?

If both are null, the question effectively changes to:
- Which operations can you handle that are not necessarily source specific? 

In the credential service, we talk in Uris, not sources, so the source concept is non-existent,s so we can't ask that question.

jainaashish commented on src/NuGet.Core/NuGet.Protocol/Plugins/PluginManager.cs:
```diff
@@ -0,0 +1,368 @@
+        /// <remarks>This is non-private only to facilitate testing.</remarks>
+        public IEnvironmentVariableReader EnvironmentVariableReader { get; private set; }
+
+        private PluginManager()
```
where are we using this private constructor?

jainaashish replied:
ignore it

jainaashish commented on src/NuGet.Core/NuGet.Protocol/Plugins/PluginManager.cs:
```diff
@@ -0,0 +1,368 @@
+        /// </summary>
+        public void Dispose()
+        {
+            if (!_isDisposed)
```
isn't this singleton and lives as long as parent process lives? what cases it can fault into multiple calls into Dispose?

nkolev92 replied:
That's correct. 

I've added a comment to explain this. 

Basically since the PluginManager owns objects which implement IDisposable and should be implementing it, we have to provide the ability here. 

However, dispose is not part of the IPluginManager interface so therefore no user should be able to dispose it.

jainaashish commented on src/NuGet.Core/NuGet.Protocol/Plugins/PluginManager.cs:
```diff
@@ -0,0 +1,368 @@
+                return string.Equals(
+                        PluginFilePath,
+                        other.PluginFilePath,
+                        StringComparison.OrdinalIgnoreCase)
```
should it OS based string comparison?

nkolev92 replied:
good point. Updated it.

jainaashish reviewed (approved):
Some comments and few changes needs to be made before merge. Rest looks good.

nkolev92 committed:
address feedback
